src/coreclr/interpreter/compileropt.cpp

```diff
@@ -0,0 +1,45 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+#include "interpreter.h"
+
+// Allocates the offset for var at the stack position identified by
+// *pPos while bumping the pointer to point to the next stack location
+int32_t InterpCompiler::AllocVarOffset(int var, int32_t *pPos)
+{
+    int32_t size, offset;
+
+    offset = *pPos;
+    size = m_pVars[var].size;
+
+    m_pVars[var].offset = offset;
+
+    *pPos = ALIGN_UP_TO(offset + size, INTERP_STACK_SLOT_SIZE);
+
+    return m_pVars[var].offset;
+}
+
+void InterpCompiler::AllocVarOffsetCB(int *pVar, void *pData)
+{
+    AllocVarOffset(*pVar, &m_totalVarsStackSize);
+}
+
+void InterpCompiler::AllocOffsets()
+{
+    // FIXME add proper offset allocator
+    InterpBasicBlock *pBB;
+
+    for (pBB = m_pEntryBB; pBB != NULL; pBB = pBB->pNextBB)
+    {
+        InterpInst *pIns;
+
+        for (pIns = pBB->pFirstIns; pIns != NULL; pIns = pIns->pNext)
+            ForEachInsSVar(pIns, NULL, &InterpCompiler::AllocVarOffsetCB);
+    }
+    m_totalVarsStackSize = ALIGN_UP_TO(m_totalVarsStackSize, INTERP_STACK_ALIGNMENT);
+    m_paramAreaOffset = m_totalVarsStackSize;
+}
+
+void InterpCompiler::OptimizeCode()
+{
+    // TODO
+}
```

Comment:
```suggestion
```
We should not be investing into code optimizations in this interpreter. Optimizing interpreter should be built from RyuJIT codebase and leverage all optimizations that RyuJIT does.

Reply:
Optimization in general for the interpreter is something far away, at least 1 year, so no work will be done on it for the prototype anyway. I think reusing RyuJIT optimizations requires a complete redesign of the interpreter code gen so it has a similar way of representing all the compilation data. Doesn't seem very attractive to me, especially given we already have a set of decently advanced optimization on mono already that can plug in very easily. So it seems like much work for little gain at this moment.

Reply:
> Doesn't seem very attractive to me, especially given we already have a set of decently advanced optimization on mono that can plug in very easily.

The primary goal of this project is to minimize our long-term costs. We do not want to be maintaining two advanced optimizers.

> I think reusing RyuJIT optimizations requires a complete redesign of the interpreter code gen so it has a similar way of representing all the compilation data.

Once we get there, we should look at this as building RyuJIT backend that produces interpreter byte code.